### PR TITLE
Private Profile Photos

### PR DESCRIPTION
**Pull request by Samyoul**
Addressing the issue of https://github.com/status-im/status/issues/56 to disable broadcasting profile images to the network by default.

The react side of https://github.com/status-im/status-go/pull/2151

Using https://github.com/status-im/status-react/pull/11650/files as a template

---

A new option in the privacy and security settings:

The default the option is "Contact"

Changing option persists

Log out

Log back in

Selection is still there.

### CODE REVIEW

File: src/status_im/ui/screens/privacy_and_security_settings/views.cljs [hunk 1]
```diff
@@ -89,3 +103,19 @@
           :on-press            #(re-frame/dispatch [:navigate-to :delete-profile])
           :accessibility-label :dapps-permissions-button
           :chevron             true}])]]))
+
+(defn radio-item [id value]
+  [quo/list-item
+   {:active    (= value id)
+    :accessory :radio
+    :title     (get titles id)
+    :on-press  #(re-frame/dispatch [:multiaccounts/profile-picture-show-to-switched id])}])
```
Comment (cammellos) [inline src/status_im/ui/screens/privacy_and_security_settings/views.cljs]: Wrong event name here:
`multiaccounts.ui/profile-picture-show-to-switched`, as you defined it in 
https://github.com/status-im/status-react/pull/11768/files#diff-c86795f915dded69102eae21e324dbf14197b79a635d3123611a7ee5ea926770R154

File: src/status_im/ui/screens/privacy_and_security_settings/views.cljs [hunk 2]
```diff
@@ -89,3 +103,19 @@
...
+
+(views/defview profile-pic-show-to []
+  (views/letsubs [{:keys [profile-pictures-show-to]} [:multiaccount/profile-pictures-show-to]]
```
Comment (cammellos) [inline src/status_im/ui/screens/privacy_and_security_settings/views.cljs]: Here  you are destructuring an object, but `[:multiaccounts/profile-pictures-show-to]` returns an integer,
so it should simply be:
```
profile-pictures-show-to [:multiaccount/profile-pictures-show-to]
```

**Comment by Samyoul** [comment]
How do I resolve the conflict while at the same time pointing to the branch that gives the functionality I need to use?

**Review by shivekkhurana — APPROVED** [review]

**Comment by cammellos** [comment]
> How do I resolve the conflict while at the same time pointing to the branch that gives the functionality I need to use?

1) Rebase your status-go against develop
2) Push changes to your branch
3) Rebase your status-react branch to status-react develop
4) update status-go version to your branch

**Comment by Samyoul** [comment]
I attempted an interactive rebase to clean this PR up but I kept getting `noop`, and no commits to target. Manually entering the commit hashes didn't work.

In the end I created a new branch `cherry-pick`ed my original commits into the new branch and switch the names of the old and new branches. Force pushed.

**Comment by cammellos** [comment]
@Samyoul would you mind rebasing status-go and status-react please? 
Ping me once done that I will put it to testing after, thanks!

**Comment by churik** [comment]
@Samyoul 
if PR is ready for manual testing, please resolve conflict and I'll relaunch e2e tests.
Thank you.

**Comment by qoqobolo** [comment]
@Samyoul Could you give an update on the current state of PR?
If it's ready to be tested, rebase it against the latest develop, please.

**Comment by Samyoul** [comment]
Hey @qoqobolo, all rebased ready to go.

**Comment by Samyoul** [comment]
@qoqobolo I notice:

> Device 1: Profile picture was not updated in chat

Is a failure.

This test may need to be updated, as this PR introduces new behaviour for profile pictures. By default profile pictures (PP) are only shown to your contacts with this PR.

So perhaps test the following:

|User Setting |Recipient |Expected behaviour|
|---|---|---|
|Show PP to only contacts |Non contact |Should not see the user's PP|
|Show PP to only contacts |Contact |Should see the user's PP|
|Show PP to everyone |Non Contact |Should see the user's PP|
|Show PP to everyone |Contact | Should see user's PP|
|Show PP to no-one |Non contact | Should not see user's PP|
|Show PP to no-one |Contact | Should not see user's PP|

**Comment by qoqobolo** [comment]
@Samyoul yeah, this test is expected to fail due to changes. So Tetiana or Sergey will update it ASAP. 
For now, we just skip it. 
Thanks for your attention to the test results, explanation, and nice table! ^_^

**Comment by qoqobolo** [comment]
@Samyoul 

### ISSUE 1: If User 1 has `See PP FROM: Everyone`, User 2 has `Show your PP TO: Everyone`, but later User 2 changes it to `Show your PP TO: Contacts`, then PP of User 1 is still will be shown to User 2.

**_Steps:_**

User 1:
- Set Profile picture
- `Privacy and security` > `Show your PP to: Everyone`
- Join a public chat
- Send a message

User 2 (not contact):
- Open `Profile` > `Appearance` > `See PP from: Everyone`
- Join the same public chat
- Check that can see PP of User 1 in the chat history

User 1:
- `Privacy and security` > `Show your PP to: Contacts/None`
- Send a message to the public chat

User 2:
- Check that  can still see PP of User 1

**Expected result: User 2 can't see PP of User 1 after settings changing**

### ISSUE 2: If User 1 adds User 2 to contacts, starts 1-1 chat, User 2 changes the PP setting to `See PP FROM: Everyone`, then changes will be applied only after User 1 and User 2 join the same public chat and User 1 sends a message. 

**_Steps:_**

User 1:
- Set Profile picture
- Add User 2 to contacts via `Profile` > `Contacts` > `+ Add new contact`
- Start 1-1 chat with User 2
- Send message

User 2:
- Open 1-1 chat with User 1 from the Activity center
- Navigate to `Profile` > `Appearance` > `See PP from: Everyone`
- Check the chat with User 1
**Expected result: can see PP of User 1**
**Actual result: can't see PP of User 1**

User 1:
- Join a public chat

User 2:
- Join the same public chat

User 1:
- Send a message to the public chat

User 2:
- Check that can see PP of User 1 now

### ISSUE 3/Question: Can we have "Show your PP to" and "See PP from" in the same setting section? 
Currently, the first is in the "Privacy and Security", and the second is in the "Appearance".
Probably, it would be better to place these two close-in-meaning settings one under the other. It would be less confusing.
cc @errorists

**Comment by errorists** [comment]
Yes, good points on the placement of setting, it should be moved out from under `Appearance`

**Comment by flexsurfer** [comment]
@Samyoul is this still relevant ?

**Comment by Samyoul** [comment]
@flexsurfer Hey, yes this is still relevant. I've just not prioritised this feature. I'll follow up with some needed changes.

@qoqobolo Hey, so:

**ISSUE 1 :**

Great find, but unfortunately this is unavoidable behaviour for the moment. We need to implement some dedicated syncing functionality before the app can reliably remove contact images. This will require a separate PR in `status-go`.

This issue happens at the moment because once a user broadcasts their profile picture we don't have reliable mechanism to know when the image should be deleted. This can be addressed with syncing account profile images so that all users will have a consistent profile image across all their devices, and contacts and correspondents can be sure that the image received or not received matches the intent of the user.

Action points for Issue 1:
- Keep current behaviour
- Implement new profile picture settings sync across user devices PR
- Implement delete contact's image if no longer decryptable

**ISSUE 2 :**

This is a bug introduced by the activity centre, I will push a fix for this.

**ISSUE 3 :**

Seems like a reasonable, I'll update with a change that moves profile pictures settings to be entirely in security and privacy.

**Comment by flexsurfer** [comment]
hey @Samyoul is this still relevant ?

**Comment by Samyoul** [comment]
Ah @flexsurfer yes it is. Let me tidy it and get it sorted for QA

**Comment by Samyoul** [comment]
@qoqobolo This feature is ready for another manual e2e . All problems should be resolved.

**Comment by churik** [comment]
@Samyoul 
Can you please rebase it? 
thanks!

**Comment by flexsurfer** [comment]
@churik done

**Comment by qoqobolo** [comment]
@Samyoul @flexsurfer thanks!

### ISSUE 4

@Samyoul unfortunately, can't see User's A profile picture on the device of UserB with default settings (See PP from: Contacts, Show your PP to: Contacts) if both users are added to contacts to each other :(

**Steps:**
1. UserA sets profile picture
2. UserA adds UserB to contacts
3. UserB adds UserA to contacts
4. UserA starts 1-1 chat with UserB

**Expected result:** UserB can see PP of UserA
**Actual result:** UserB does not see PP

**Comment by cammellos** [comment]
@qoqobolo would you mind testing again, the issue should be fixed? Thank you

**Comment by qoqobolo** [comment]
@cammellos thanks for the fix!
But well, there is still pretty strange behavior here.

Currently, ISSUE 4 is reproducible sporadically.
In 4 out of 5 cases, the feature worked for me if the user with the profile picture (UserA) added the second user (UserB) to the contacts **_first_**. In that case, UserB immediately saw the PP after adding UserA to contacts. 


I mean these **steps to reproduce successful flow**:
1. UserA sets profile picture
2. UserA adds UserB to contacts (e.g via QR scanner)
3. UserB adds UserA to contacts (via QR scanner)
4. UserA starts 1-1 chat with UserB

But it doesn't work for me anymore :(

However, if after that both users join the same public chat, and UserA (with PP) sends a message, then the PP will be displayed for UserB in the chat history and in 1-1 chat as well.

So, this issue can be described:

#### ISSUE 5: Profile photo is only visible for UserB after UserA sends a message to public chat

**Steps:**
1. UserA sets profile picture
2. UserA adds UserB to contacts (e.g via QR scanner)
3. UserB adds UserA to contacts (via QR scanner)
4. UserA starts 1-1 chat with UserB
5. UserA and UserB join the same public chat
6. UserA sends a message

**Expected result:** UserB can see PP after adding UserA to contacts in User Profile and in 1-1 chat
**Actual result:** UserB can only see UserA's PP after joining a public chat and sending message by UserA

**Comment by cammellos** [comment]
@qoqobolo thanks for the info, I will be taking a look at it

**Comment by qoqobolo** [comment]
@cammellos issues 4 and 5 are fixed 🥳 Default settings `See PP from: Contacts`, `Show your PP to: Contacts` work fine. Thanks! 

#### ISSUE 6: Profile photo of UserA is visible to UserB if UserA sets `Show your PP to: None` or `Show your PP to: Contacts`, but UserB is not contact of UserA

The issue is reproducible only if UserB has already seen UserA's photo profile, i.e if at first UserA's had `Show your PP to: Contacts` and UserB were contact of UserA, but then UserA changed this setting to `Show your PP to: None`. 
So if UserA first changed the settings to `Show your PP to: None`, and then started a chat with UserB, then UserB will not see UserA's profile photo.

**Reproduction:**
1. Add profile photo
2. Add UserB to contacts and start 1-1 chat
3. UserB adds UserA to contacts
4. UserA: Settings > Privacy and security > `Show your PP to: None`
5. UserB: check profile photo of UserA
**Expected behavior:** UserB **cannot** see PP of UserA
**Actual behavior:**  UserB sees PP of UserA

There are cases when PP is visible, although it should be hidden.

1. UserA (with PP) sets `Show your PP to: Contacts`, UserB `See PP from: Everyone`
UserB is not contact of UserA

2. UserA (with PP) sets `Show your PP to: None`, UserB `See PP from: Everyone`

3. UserA (with PP) sets `Show your PP to: None`, UserB `See PP from: Contacts`
UserB adds to contacts UserA


For all these cases
**Expected behavior:** UserB **cannot** see PP of UserA
**Actual behavior:**  UserB sees PP of UserA

Please ping me for clarification if needed.

**Comment by churik** [comment]
@cammellos I wouldn't mind merging it with https://github.com/status-im/status-react/pull/11768#issuecomment-931178854 but if it is too much effort to fix it now, as a temporary decision might be for example to warn the user that he couldn't hide PP anymore for users that already seen his PP (add some explanation to `Show my picture to` screen)
Can we do it in this PR or we should log separate issue?

**Comment by cammellos** [comment]
@churik if you don't mind logging on a separate issue so we can get this merged, and we can add it to 1.17 so we make sure we don't release without, if that's ok

**Comment by churik** [comment]
Sure, we will do this
Thank you @qoqobolo for testing and @cammellos for resolving issues!

**Comment by cammellos** [comment]
Thank you @churik @qoqobolo @Samyoul , merged!

**Pull request merged**